File: DraftRightWindows/DraftRightWindows/Models/TriggerMode.cs
namespace DraftRightWindows.Models;

/// <summary>
/// How a rewrite is triggered — the pencil button on selection, or a global
/// hotkey. Mutually exclusive: exactly one is active at a time. Single source of
/// truth, decoupled from whether a hotkey combo is set (so switching to the
/// pencil keeps the saved shortcut). Mirrors the macOS <c>TriggerMode</c>
/// (DraftRight #180).
/// </summary>
public enum TriggerMode
{
    /// <summary>The pencil button appears when text is highlighted by dragging.</summary>
    Pencil,

    /// <summary>A global keyboard shortcut opens the panel.</summary>
    Hotkey
}

public static class TriggerModeExtensions
{
    /// <summary>Whether the on-selection pencil should be active in this mode.</summary>
    public static bool UsesPencil(this TriggerMode mode) => mode == TriggerMode.Pencil;

    /// <summary>Whether the global hotkey should be registered in this mode.</summary>
    public static bool UsesHotkey(this TriggerMode mode) => mode == TriggerMode.Hotkey;

    /// <summary>
    /// Stable lowercase string used for JSON persistence. MUST stay byte-identical
    /// to macOS <c>TriggerMode.rawValue</c> and the Linux equivalent so the three
    /// platforms' <c>settings.json</c> stays interoperable (same discipline as
    /// <see cref="AppModeExtensions.ApiValue"/>).
    /// </summary>
    public static string ApiValue(this TriggerMode mode) => mode switch
    {
        TriggerMode.Pencil => "pencil",
        TriggerMode.Hotkey => "hotkey",
        _ => "hotkey"
    };

    public static TriggerMode FromApiValue(string? raw) => raw switch
    {
        "pencil" => TriggerMode.Pencil,
        "hotkey" => TriggerMode.Hotkey,
        // "both" was removed (#180); a stored value migrates to the hotkey default.
        _ => TriggerMode.Hotkey
    };

    public static string DisplayName(this TriggerMode mode) => mode switch
    {
        TriggerMode.Pencil => "Pencil",
        TriggerMode.Hotkey => "Hotkey",
        _ => "Hotkey"
    };
}

File: DraftRight/TriggerMode.swift
import Foundation

/// How the rewrite panel is triggered.
///
/// This is the single source of truth for which trigger mechanisms are active,
/// decoupled from whether a hotkey combo happens to be set. Previously the mode
/// was inferred from `hotkeyString` being empty or not — a boolean, so only two
/// states were possible. An explicit enum adds the third (`both`) and keeps the
/// hotkey combo as pure data that survives a mode switch (#179).
enum TriggerMode: String, CaseIterable, Identifiable {
    /// The pencil button appears when text is selected.
    case pencil
    /// A global keyboard shortcut opens the panel.
    case hotkey
    /// Both at once — the pencil appears AND the shortcut fires.
    case both

    var id: String { rawValue }

    /// Whether the on-selection pencil button should be active in this mode.
    var usesPencil: Bool { self == .pencil || self == .both }

    /// Whether the global hotkey should be registered in this mode.
    var usesHotkey: Bool { self == .hotkey || self == .both }

    var displayName: String {
        switch self {
        case .pencil: return "Pencil"
        case .hotkey: return "Hotkey"
        case .both: return "Both"
        }
    }
}
